reference/JSDoc-Type-References.md

```diff
@@ -0,0 +1,242 @@
+which the compiler takes as a signal that the code is being written
+with Typescript in mind. Old versions of Typescript also rewrote
+`object -> any`, but stopped in Typescript 3.7.
+
+There are also rewrites of other JSDoc types that are equivalent to
+built-in Typescript types:
+
+Type                  | Resolved Type
+----------------------|--------------
+`Null`                | `null`
+`Undefined`           | `undefined`
+`Void`                | `void`
+`function`            | `Function`
+`Object<string,Type>` | `{ [x: string]: Type }`
+`Object<number,Type>` | `{ [x: number]: Type }`
+
+Finally, `Array` and `Promise` both default their type parameters to
+`any` when no type parameters are provided:
+
+Type        | Resolved Type
+------------|--------------
+`Array`     | `any[]`
+`Promise`   | `Promise<any>`
+
+This is standard for all type references in JSDoc, so I don't think
+it's needed anymore. It only applies when `"noImplicitAny": false`.
+
+### Where To Find The Code ###
+
+`getIntendedJSDocTypeReference` in src/compiler/checker.ts.
+
+## Values as Types ##
+
+The compiler maintains three namespaces: one each for values, types and
+namespaces. Put another way, each name can have a value meaning, a
+type meaning and a namespace meaning. However, there are fewer ways
+to declare and reference types in Javascript, and the distinction
+between type and value can be subtle, so for Javascript, the compiler
+tries to resolve type references as values if it can't find a type.
+
+Here's an example:
+
+```ts
+const FOO = "foo"
+const BAR = "bar"
+
+/** @param {FOO | BAR} type */
+function f(type) {
+}
+```
+
+In the compiler, this is resolved in two stages. Both stages use a
+JSDoc fallback. First the compiler resolves the name; then it finds
+the type for the name. Normally type resolution only looks in the type
+namespace, but when it finds nothing there, a fallback looks in the
+value namespace.
+
+In the example above, `FOO` is a value, a block-scoped `const`. During
+type resolution, there is no type named `FOO`. So instead the compiler
+uses the value named `FOO`.
+
+Now there is a problem: the value `FOO` has no type declaration, only
+a value declaration. So we now get the type of the value declaration
+instead of the type of the type declaration.
+
+I think it would help to look at an example
+that has both a value declaration and a type declaration:
+
+```ts
+var i = 0
+interface i {
+  e: 1
+  m: 1
+}
+```
+
+If you ask for the type of the value declaration of `i`, you'll get
+`number`. If you ask for the type of type declaration of `i`, you'll
+get `{ e: 1, m: 1 }`. In Javascript, you can't *write* `interface`, so
+it would just be
+
+```js
+var i = 0
+```
+
+But what if you treat `i` as a type?
+
+```js
+/** @type {i} */
+var j
+```
+
+The compiler just reuses the type of the value declaration: `number`.
+This is equivalent to inserting the `typeof` type operator to get
+`typeof i`.
```

Comment:
The existing handbook page doesn't really talk about type resolution, hoping that we support existing patterns well enough that people don't have to think about it. And as @weswigham points out, it's something that we don't want to talk about too much in case we find a better way to accomplish the same thing.

Reply:
Yes, I see the point in what he says.  To be clear, I read this whole thing from an "internals documentation" POV rather than a user's.  (The existing jsdoc page looked to me like the best user's view description, and I've assumed that it would be ported..?)